src/sidebar/store/modules/drafts.js

```diff
@@ -0,0 +1,194 @@
+  match(annotation) {
+    return (
+      (this.annotation.$tag && annotation.$tag === this.annotation.$tag) ||
+      (this.annotation.id && annotation.id === this.annotation.id)
+    );
+  }
+  /**
+   * Return true if this draft is empty and can be discarded without losing
+   * any user input.
+   */
+  isEmpty() {
+    return !this.text && this.tags.length === 0;
+  }
+}
+
+/* Reducer */
+
+const update = {
+  DISCARD_ALL_DRAFTS: function() {
+    return {
+      drafts: [],
+    };
+  },
+  REMOVE_DRAFT: function(state, action) {
+    const drafts = state.drafts.filter(draft => {
+      return !draft.match(action.annotation);
+    });
+    return {
+      drafts,
+    };
+  },
+  UPDATE_DRAFT: function(state, action) {
+    // removes a matching existing draft, then adds
+    const drafts = state.drafts.filter(draft => {
+      return !draft.match(action.draft.annotation);
+    });
+    drafts.push(action.draft); // push ok since its a copy
+    return {
+      drafts,
+    };
+  },
+  ADD_DRAFT: function(state, action) {
+    return {
+      drafts: state.drafts.concat(action.draft), // non-mutating
+    };
+  },
+};
+
+const actions = util.actionTypes(update);
+
+/* Actions */
+
+/**
+ * Create or update the draft version for a given annotation by
+ * replacing any existing draft or simply creating a new one.
+ */
+function createDraft(annotation, changes) {
+  return {
+    type: actions.UPDATE_DRAFT,
+    draft: new Draft(annotation, changes),
+  };
+}
+
+/** Remove all drafts. */
+function discardAllDrafts() {
+  return {
+    type: actions.DISCARD_ALL_DRAFTS,
+  };
+}
+
+/** Remove the draft version of an annotation. */
+function removeDraft(annotation) {
+  return {
+    type: actions.REMOVE_DRAFT,
+    annotation,
+  };
+}
+
+/* Selectors */
+
+/**
+ * Returns the number of drafts - both unsaved new annotations, and unsaved
+ * edits to saved annotations - currently stored.
+ *
+ * @return {number}
+ */
+function countDrafts(state) {
+  return state.drafts.length;
+}
+
+/**
+ * Retrieve the draft changes for an annotation.
+ *
+ * @return {Draft|null}
+ */
+function getDraft(state, annotation) {
+  for (let i = 0; i < state.drafts.length; i++) {
+    const draft = state.drafts[i];
+    if (draft.match(annotation)) {
+      return draft;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Returns the draft changes for an annotation, or null if no draft exists
+ * or the draft is empty.
+ *
+ * Since this is a non-primitive object injected into the store, there are
+ * some consequences with serialization. Slack discussion:
+ * https://hypothes-is.slack.com/archives/C1M8NH76X/p1562779754051700?thread_ts=1562776681.050500&cid=C1M8NH76X
+ *
+ * @return {Draft|null}
+ */
+function getDraftIfNotEmpty(state, annotation) {
+  const draft = getDraft(state, annotation);
+  if (!draft) {
+    return null;
+  }
+  return draft.isEmpty() ? null : draft;
+}
+
+/**
+ * Returns a list of draft annotations which have no id.
+ *
+ * @return {Object[]}
```

Comment:
```suggestion
 * @return {Draft[]}
```

Reply:
@robertknight `unsavedDrafts` returns an array of annotation objects, but not the entire Draft itself. Perhaps we should change the name of the selector to `unsavedAnnotations`

?

Reply:
Ah, of course. I was just looking at the function name which had confused me. `unsavedAnnotations` is a better name IMO. I notice that the single consumer in `hypothesis-app.js` has an incorrect or misleading parameter name in its callback.